### PR TITLE
Core: Add JSONMessagePart for Hint Status (Hint Priority)

### PR DESCRIPTION
**Pull request by CodeTriangle**
## What is this fixing or adding?

Added a new JSONMessagePart type (`hint_status`) for displaying hint status/hint priority, pursuant to [my suggestion here](https://discord.com/channels/731205301247803413/731214280439103580/1320078772199882864) to fix [this issue](https://discord.com/channels/731205301247803413/731214280439103580/1320009008156901449).

This sets up correct handlers throughout the code to display hint priority in the correct coloration and to send abstract styling hints to clients.

### Some design notes

A correctly formatted hint_status JSONMessagePart appears as:

```json
{"text": "(unspecified)", "hint_status": 1, "type": "hint_status"}
```

In contrast to `item_id` and `player_id`, I elected to store hint status and text separately as there will not be one canonical visual representation of the text. I could be convinced that this system should be changed. 

## How was this tested?

Poked around in CommonClient and tested different values of hint priority.

## If this makes graphical changes, please attach screenshots.

This is a more versatile means to the same visual end as #3506 added. I believe parity has been achieved.

### CODE REVIEW

File: docs/network protocol.md [hunk 1]
```diff
@@ -554,6 +554,7 @@ class JSONMessagePart(TypedDict):
     color: Optional[str] # only available if type is a color
     flags: Optional[int] # only available if type is an item_id or item_name
     player: Optional[int] # only available if type is either item or location
+    hint_status: Optional[int] # only available if type is hint_status
```
Comment (Jarno458) [inline docs/network protocol.md:557]: I guess this should be Optional[HintStatus] rather than int

**Review by Jarno458 — CHANGES_REQUESTED** [review]

File: docs/network protocol.md (hunk 1 above)
Comment (EmilyV99) [reply docs/network protocol.md:557]:
```suggestion
    hint_status: Optional[[HintStatus](#HintStatus)] # only available if type is hint_status
```
aye, that would seem more correct.

File: docs/network protocol.md [hunk 2]
```diff
@@ -569,6 +570,7 @@ Possible values for `type` include:
 | location_id | Location ID, should be resolved to Location Name |
 | location_name | Location Name, not currently used over network, but supported by reference Clients. |
 | entrance_name | Entrance Name. No ID mapping exists. |
+| hint_status | The status of the hint (i.e. found, priority). See [HintStatus](#HintStatus) for details. Both `text` and `hint_status` are given. |
```
Comment (EmilyV99) [inline docs/network protocol.md:573]:
```suggestion
| hint_status | The status of the hint (i.e. found, priority). Both `text` and `hint_status` are given. |
```
Don't need this bit here, if you apply my other suggestion as well.

**Review by EmilyV99 — CHANGES_REQUESTED** [review]
Haven't tested the code atm, but don't *see* anything wrong with it. Just a couple notes on the documentation.

**Comment by CodeTriangle** [comment]
Thanks for the suggestions; they have been added.

File: docs/network protocol.md [hunk 3]
```diff
@@ -554,6 +554,7 @@ class JSONMessagePart(TypedDict):
     color: Optional[str] # only available if type is a color
     flags: Optional[int] # only available if type is an item_id or item_name
     player: Optional[int] # only available if type is either item or location
+    hint_status: Optional[[HintStatus](#HintStatus)] # only available if type is hint_status
```
Comment (EmilyV99) [inline docs/network protocol.md:557]:
```suggestion
    hint_status: Optional[HintStatus] # only available if type is hint_status
```

File: docs/network protocol.md [hunk 4]
```diff
@@ -569,6 +570,7 @@ Possible values for `type` include:
 | location_id | Location ID, should be resolved to Location Name |
 | location_name | Location Name, not currently used over network, but supported by reference Clients. |
 | entrance_name | Entrance Name. No ID mapping exists. |
+| hint_status | The status of the hint (i.e. found, priority). Both `text` and `hint_status` are given. |
```
Comment (EmilyV99) [inline docs/network protocol.md:573]:
```suggestion
| hint_status | The [HintStatus](#HintStatus) of the hint. Both `text` and `hint_status` are given. |
```

**Review by EmilyV99 — CHANGES_REQUESTED** [review]
Oh... oops. No way to preview the md rich diff when making suggestions in review in the web editor is really annoying, and made me miss that this is in a code block, so that does NOT work. So uh, these will both need to change a bit more...

**Comment by CodeTriangle** [comment]
> Oh... oops. No way to preview the md rich diff when making suggestions in review in the web editor is really annoying, and made me miss that this is in a code block, so that does NOT work. So uh, these will both need to change a bit more...

Good catch. Committed and double checked the formatting myself this time. Looks good.

As a tangent, I'd welcome suggestions to improve the documentation of this new feature in general. I was working with somewhat limited time and laptop battery, so I'm not too confident in how it turned out.

**Review by EmilyV99 — APPROVED** [review]
> As a tangent, I'd welcome suggestions to improve the documentation of this new feature in general. I was working with somewhat limited time and laptop battery, so I'm not too confident in how it turned out.

Looks straightforward enough to me.

Just tested this by updating GodotAP to handle this new message syntax; works precisely as I would expect (both from what the docs say, and from what I'd want this to accomplish). (see below; I map the received hint_status value to the appropriate plum color for priority, color entirely handled on MY end instead of server end)
image

Also tested TextClient, which is showing the color correctly.
image

So, all looks good!

**Comment by Jarno458** [comment]
I tested today aswel it seems to work
```
[{"cmd":"PrintJSON","data":[{"text":"[Hint]: "},{"text":"1","type":"player_id"},{"text":"'s "},{"text":"1337116","player":1,"flags":1,"type":"item_id"},{"text":" is at "},{"text":"1337109","player":2,"type":"location_id"},{"text":" in "},{"text":"2","type":"player_id"},{"text":"'s World"},{"text":". "},{"text":"(priority)","hint_status":30,"type":"hint_status"}],"type":"Hint","receiving":1,"item":{"item":1337116,"location":1337109,"player":2,"flags":1,"class":"NetworkItem"},"found":false}]
```

**Review by Jarno458 — APPROVED** [review]

**Review by NewSoupVi — APPROVED** [review]

**Pull request merged**